zkvm/src/types.rs

```diff
@@ -301,11 +341,73 @@ impl Value {
 
 impl Expression {
     pub fn constant<S: Into<Scalar>>(a: S) -> Self {
-        let a: Scalar = a.into();
+        Expression::Constant(ScalarWitness::Scalar(a.into()))
+    }
+}
+
+impl Neg for Expression {
+    type Output = Expression;
+
+    fn neg(self) -> Expression {
+        match self {
+            Expression::Constant(a) => Expression::Constant(-a),
+            Expression::Terms(mut terms, assignment) => {
+                for (_, n) in terms.iter_mut() {
+                    *n = -*n;
+                }
+
+                let x = match assignment {
+                    Some(a) => Some(-a),
+                    None => None,
+                };
+                Expression::Terms(terms, x)
+            }
+        }
+    }
+}
+
+impl Add for Expression {
+    type Output = Expression;
+
+    fn add(self, rhs: Expression) -> Expression {
+        match (self, rhs) {
+            (Expression::Constant(a), Expression::Constant(b)) => Expression::Constant(a + b),
+            (Expression::Constant(a), Expression::Terms(mut terms, assignment)) => {
```

Comment:
let's make it clearer (and similarly for the alternative branch):
```suggestion
            (Expression::Constant(left), Expression::Terms(mut right_terms, right_assignment)) => {
```